pkgs/applications/science/molecular-dynamics/lammps/default.nix

```diff
@@ -35,6 +38,10 @@
   SRD = true;
   REAXFF = true;
 }
+# Extra cmakeFlags to add as "-D${attr}=${value}"
+, extraCmakeFlags ? {}
+# Extra `buildInputs` - meant for packages that require more inputs
+, extraBuildInputs ? []
```

Comment:
> Is there a particular motivation why you want to add MPI in that fashion, and deviate from the widely used 'withMpi' pattern?

I just don't see a reason why we would treat `mpi` differently then other "packages" and optional dependencies for this package. I want all of the switches and knobs of this package to have the same interface.

> It requires, for example, additional binaries from the mpi package, such as mpirun.

I understand. If people use `lammps` in their development environment, now they can simply add `lammps-mpi.passthru.extraBuildInputs` to their devShell's `buildInputs`, and get the same `mpi` as that of `lammps-mpi`.

Reply:
> I just don't see a reason why we would treat mpi differently then other "packages" and optional dependencies for this package. I want all of the switches and knobs of this package to have the same interface.

It's probably a matter of taste how special mpi is. I would argue that MPI changes the runtime environment, which requires extra tools to even run the binary and thus has a special standing amongst the inputs. 

> I understand. If people use lammps in their development environment, now they can simply add lammps-mpi.passthru.extraBuildInputs to their devShell's buildInputs, and get the same mpi as that of lammps-mpi.

Yes, that would work. When mpi is in `propagatedBuildInputs` it would even work without special intervention.

Using `extraBuildInputs` is a bit of a paradigm shift, that is not widely used (but an interesting approach). Wouldn't it be better to have uniform solution?

Reply:
> Using `extraBuildInputs` is a bit of a paradigm shift, that is not widely used (but an interesting approach). Wouldn't it be better to have uniform solution?

What kind of a solution would you like to see? I mean, unifying which ideas?

> Yes, that would work. When mpi is in `propagatedBuildInputs` it would even work without special intervention.

We could also add `propagatedBuildInputs` as an argument to the expression... 

With a good message in `passthru.mpi = throw "...";` and a release notes entry, perhaps this is good enough for `nixos-unstable`?

TBH I feel very comfortable introducing breaking changes to the interface of this package, as it was very outdated no long ago, and I doubt there were many if any, users relying on the `passthru.mpi` interface.